File: apps/starter-ios/StarterApp/Core/Billing/BillingService.swift
import Foundation

struct BillingService {
    func fetchEntitlements() async throws -> EntitlementState {
        try await APIClient.shared.request("billing/entitlements")
    }

    func verifyNativePurchase(
        productId: String,
        purchaseToken: String,
        transactionId: String? = nil
    ) async throws -> VerifyNativePurchaseResponse {
        try await APIClient.shared.request(
            "billing/verify/native",
            method: "POST",
            body: VerifyNativePurchaseBody(
                platform: "ios",
                productId: productId,
                purchaseToken: purchaseToken,
                transactionId: transactionId
            )
        )
    }
}

File: apps/starter-api/src/billing/billing.service.ts
import { Injectable } from '@nestjs/common';
import { VerifyNativePurchaseDto } from './dtos/verify-native-purchase.dto';

type EntitlementState = {
  tier: 'free' | 'tracker';
  source: 'starter' | 'ios' | 'android';
  features: string[];
  renewsAt: string | null;
};

@Injectable()
export class BillingService {
  private readonly entitlements = new Map<string, EntitlementState>();

  getEntitlements(userId: string) {
    return (
      this.entitlements.get(userId) ?? {
        tier: 'free',
        source: 'starter',
        features: ['basic_logging', 'exercise_library'],
        renewsAt: null,
      }
    );
  }

  verifyNativePurchase(userId: string, dto: VerifyNativePurchaseDto) {
    const nextState: EntitlementState = {
      tier: 'tracker',
      source: dto.platform,
      features: [
        'basic_logging',
        'exercise_library',
        'unlimited_templates',
        'advanced_analytics',
        'export_data',
      ],
      renewsAt: new Date(Date.now() + 30 * 24 * 60 * 60 * 1000).toISOString(),
    };

    this.entitlements.set(userId, nextState);

    return {
      verified: true,
      productId: dto.productId,
      entitlement: nextState,
    };
  }
}
